flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapKeyedStateBackendBuilder.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * @param <K> The data type that the key serializer serializes.
+ */
+public class HeapKeyedStateBackendBuilder<K> extends AbstractKeyedStateBackendBuilder<K> {
+	/**
+	 * The configuration of local recovery.
+	 */
+	private final LocalRecoveryConfig localRecoveryConfig;
+	/**
+	 * Factory for state that is organized as priority queue.
+	 */
+	private final HeapPriorityQueueSetFactory priorityQueueSetFactory;
+	/**
+	 * Whether asynchronous snapshot is enabled.
+	 */
+	private final boolean asynchronousSnapshots;
+
+
+	public HeapKeyedStateBackendBuilder(
+		TaskKvStateRegistry kvStateRegistry,
+		TypeSerializer<K> keySerializer,
+		ClassLoader userCodeClassLoader,
+		int numberOfKeyGroups,
+		KeyGroupRange keyGroupRange,
+		ExecutionConfig executionConfig,
+		TtlTimeProvider ttlTimeProvider,
+		@Nonnull Collection<KeyedStateHandle> stateHandles,
+		StreamCompressionDecorator keyGroupCompressionDecorator,
+		LocalRecoveryConfig localRecoveryConfig,
+		HeapPriorityQueueSetFactory priorityQueueSetFactory,
+		boolean asynchronousSnapshots,
+		CloseableRegistry cancelStreamRegistry) {
+		super(
+			kvStateRegistry,
+			keySerializer,
+			userCodeClassLoader,
+			numberOfKeyGroups,
+			keyGroupRange,
+			executionConfig,
+			ttlTimeProvider,
+			stateHandles,
+			keyGroupCompressionDecorator,
+			cancelStreamRegistry);
+		this.localRecoveryConfig = localRecoveryConfig;
+		this.priorityQueueSetFactory = priorityQueueSetFactory;
+		this.asynchronousSnapshots = asynchronousSnapshots;
+	}
+
+	@Override
+	public HeapKeyedStateBackend<K> build() throws BackendBuildingException {
+		// Map of registered Key/Value states
+		Map<String, StateTable<K, ?, ?>> registeredKVStates = new HashMap<>();
+		// Map of registered priority queue set states
+		Map<String, HeapPriorityQueueSnapshotRestoreWrapper> registeredPQStates = new HashMap<>();
+		CloseableRegistry cancelStreamRegistryForBackend = new CloseableRegistry();
+		HeapSnapshotStrategy<K> snapshotStrategy = initSnapshotStrategy(
+			asynchronousSnapshots, registeredKVStates, registeredPQStates, cancelStreamRegistryForBackend);
+		HeapKeyedStateBackend<K> backend = new HeapKeyedStateBackend<>(
+			kvStateRegistry,
+			keySerializerProvider,
+			userCodeClassLoader,
+			numberOfKeyGroups,
+			keyGroupRange,
+			executionConfig,
+			ttlTimeProvider,
+			cancelStreamRegistryForBackend,
+			keyGroupCompressionDecorator,
+			registeredKVStates,
+			registeredPQStates,
+			localRecoveryConfig,
+			priorityQueueSetFactory,
+			snapshotStrategy
+		);
+		HeapRestoreOperation<K> restoreOperation = new HeapRestoreOperation<>(
+			restoreStateHandles,
+			keySerializerProvider,
+			userCodeClassLoader,
+			registeredKVStates,
+			registeredPQStates,
+			cancelStreamRegistry,
+			priorityQueueSetFactory,
+			keyGroupRange,
+			numberOfKeyGroups,
+			snapshotStrategy,
+			backend);
+		try {
+			restoreOperation.restore();
+		} catch (Throwable e) {
```

Comment:
Is it on purpose to catch `Throwable` instead of `Exception`? The method only declares exception and I am not sure why we would also like to catch errors etc. here.

Reply:
With `Throwable` we could catch some uncaught exception (such as NPE) and also get it explained, and also to keep in accordance with the try-catch clause in `RocksDBKeyedStateBackendBuilder`.

Reply:
I don't understand. NPEs would also be caught if we do `catch(Exception e)`.

Reply:
Correct, then no special reason if we don't need to handle errors like `AssertionError` I guess.